ksql-engine/src/main/java/io/confluent/ksql/engine/TopicAccessValidator.java

```diff
@@ -0,0 +1,169 @@
+import io.confluent.ksql.parser.tree.InsertValues;
+import io.confluent.ksql.parser.tree.Statement;
+import io.confluent.ksql.planner.plan.DataSourceNode;
+import io.confluent.ksql.planner.plan.PlanNode;
+import io.confluent.ksql.services.ServiceContext;
+import io.confluent.ksql.util.KsqlTopicAccessException;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Set;
+
+import org.apache.kafka.clients.admin.TopicDescription;
+import org.apache.kafka.common.acl.AclOperation;
+
+/**
+ * Wraps a {@link ServiceContext} to validate access to Kafka topics.
+ */
+public final class TopicAccessValidator {
+  private final ServiceContext serviceContext;
+
+  private TopicAccessValidator(final ServiceContext serviceContext) {
+    this.serviceContext = serviceContext;
+  }
+
+  /**
+   * Creates a {@code TopicAccessValidator} that uses the specified {@code ServiceContext}
+   * to check Kafka topic access.
+   *
+   * @param serviceContext The {@code ServiceContext} to use.
+   * @return A {@code TopicAccessValidator} object.
+   */
+  public static TopicAccessValidator as(final ServiceContext serviceContext) {
+    return new TopicAccessValidator(serviceContext);
+  }
+
+  /**
+   * Verifies if all query source topics have read access.
+   *
+   * @param node The node generated by the {@code QueryEngine#buildLogicalPlan}.
+   * @throws KsqlTopicAccessException If a topic does not exist or does not have permissions.
+   */
+  public void verifyQuerySourcesPermissions(final PlanNode node) throws KsqlTopicAccessException {
+    final Set<String> kafkaTopics = new HashSet<>();
+    collectKafkaTopics(kafkaTopics, node);
+
+    for (String kafkaTopic : kafkaTopics) {
+      checkAccess(kafkaTopic, Collections.singleton(AclOperation.READ));
+    }
+  }
+
+  /**
+   * Verifies if a non-query KSQL statement has permissions to access the required topics.
+   *
+   * @param statement The statement to verify for permissions.
+   * @throws KsqlTopicAccessException If a topic does not exist or does not have permissions.
+   */
+  public void verifyStatementPermissions(final Statement statement)
+      throws KsqlTopicAccessException {
+    if (statement instanceof CreateSource) {
+      verifyCreateSourcePermissions((CreateSource) statement);
+    } else if (statement instanceof InsertInto) {
+      verifyInsertIntoPermissions((InsertInto) statement);
+    } else if (statement instanceof InsertValues) {
+      verifyInsertValuesPermissions((InsertValues) statement);
+    } else if (statement instanceof CreateAsSelect) {
+      verifyCreateAsSelectPermissions((CreateAsSelect) statement);
+    }
+  }
+
+  private void verifyCreateAsSelectPermissions(final CreateAsSelect createAsSelect) {
+    /*
+     * Check permissions for CREATE STREAM and CREATE TABLE statements that have
+     * a persistent query in it.
+     *
+     * This only validates the user can CREATE and WRITE the target topic. The persistent query
```

Comment:
shouldn't we also verify READ permissions from the source topic? Otherwise you can pipe data from one topic into another and just read the sink topic

Reply:
The checks for source topics happen when the logical planner is created. That way I could get the source topic information.

Reply:
ack